src/interface.jl

```diff
@@ -11,27 +11,66 @@ function state(o, x)
   end
 end
 
-function _update(o, st, x, x̄s...)
-  st′, x̄′ = apply(o, st, x, x̄s...)
-  return st′, patch(x, x̄′)
+function _update(o, st, x, x̄s...; copy::Bool = true)
+  st′, x̄′ = apply(o, st, x, x̄s...; copy)
+  x′ = copy ? patch(x, x̄′) : patch!!(x, x̄′)
+  return st′, x′
 end
 
-function update(o, state, x::T, x̄s...) where T
+function update(o, state, x::T, x̄s...; copy::Bool = true) where T
   if all(isnothing, x̄s)
     return state, x
   elseif isnumeric(x)
-    return _update(o, state, x, x̄s...)
+    return _update(o, state, x, x̄s...; copy)
   else
     x̄s′ = map(x̄ -> functor(typeof(x), x̄)[1], x̄s)
```

Comment:
```suggestion
    x̄s′ = map(x̄ -> functor(T, x̄)[1], x̄s)
```
Or remove the type param entirely

Reply:
Oh right. Actually that's one of the lines I'm still not sure I understand. (Because exactly what `@functor` defines is cleverly hidden within a function which calls `eval`...)

Reply:
Unfortunately the macro expansion is obscured behind an intermediate function, but the translation is pretty straightforward:
```julia
struct Foo
  a
  b
  c
end
@functor Foo (a, c)

# becomes
Functors.functor(::Type{Foo}, x) = (a=x.a, c=x.c), y -> Foo(y[1], b=x.b, y[2])
```
If there's a cleaner and clearer way to pull this off, we should definitely do it.

Reply:
Ok. This whole recursion isn't quite right yet, fails on the Resnet example from the docs, but I haven't isolated why yet.